projects/RabbitMQ.Client/client/impl/MainSession.cs

```diff
@@ -44,50 +46,42 @@ namespace RabbitMQ.Client.Impl
     internal sealed class MainSession : Session
     {
         private readonly object _closingLock = new object();
-
-        private readonly ProtocolCommandId _closeProtocolId;
-        private readonly ProtocolCommandId _closeOkProtocolId;
-
-        private bool _closeServerInitiated;
-        private bool _closing;
+        private volatile bool _closeServerInitiated;
+        private volatile bool _closing;
 
         public MainSession(Connection connection) : base(connection, 0)
         {
-            connection.Protocol.CreateConnectionClose(0, string.Empty, out OutgoingCommand request, out _closeOkProtocolId);
-            _closeProtocolId = request.Method.ProtocolCommandId;
         }
 
         public Action Handler { get; set; }
 
         public override bool HandleFrame(in InboundFrame frame)
         {
-            lock (_closingLock)
-            {
-                if (!_closing)
-                {
-                    return base.HandleFrame(in frame);
-                }
-            }
-
-            if (!_closeServerInitiated && frame.Type == FrameType.FrameMethod)
+            if (_closing)
```

Comment:
can we switch the common case first?

Reply:
That'll involve doing a lot more checks really. This just does the first check on `if _closing` which is almost always false and then just continues to the transmit. This usually compiles to no code jumping unless the if matches which is usually faster.

Reply:
I expect this to be cheaper than acquiring a lock either way ;)

Reply:
This for sure =) 
A quick "is null" check might be as fast as a enum compare 🤔

Reply:
> This looks reasonable, thank you. Any benchmarks that demonstrate the improvement for some key workloads?

I'll see if I can do some benchmarks. They are a bit harder to do here since they require actual usages of Connection/Model objects.